README.md
pydf
====

[![codecov.io](https://codecov.io/github/tutorcruncher/pydf/coverage.svg?branch=master)](https://codecov.io/github/tutorcruncher/pydf?branch=master)
[![Build Status](https://travis-ci.org/tutorcruncher/pydf.svg?branch=master)](https://travis-ci.org/tutorcruncher/pydf)
[![PyPI](https://img.shields.io/pypi/v/python-pdf.svg)](https://pypi.python.org/pypi/python-pdf)

PDF generation in python using [wkhtmltopdf](http://wkhtmltopdf.org/).

Wkhtmltopdf binaries are precompiled and included in the package making pydf easier to use,
in particular this means pydf works on heroku.

Based on [pywkher](https://github.com/jwmayfield/pywkher) but significantly extended.

Currently using **wkhtmltopdf v. 0.12.2** (beta).

## Install

    pip install python-pdf

(pydf was taken, but I guess python-pdf is a clearer name anyway.)

## Basic Usage

```python
import pydf
pdf = pydf.generate_pdf('<h1>this is html</h1>')
open('test_doc.pdf', 'w').write(pdf)

pdf = pydf.generate_pdf('www.google.com')
open('google.pdf', 'w').write(pdf)
```

## API

#### generate_pdf(source, [**kwrags])

Generate a pdf from either a url or a html string.

After the html and url arguments all other arguments are
passed straight to wkhtmltopdf

For details on extra arguments see the output of get_help()
and get_extended_help()

All arguments whether specified or caught with extra_kwargs are converted
to command line args with `'--' + original_name.replace('_', '-')`.

Arguments which are True are passed with no value eg. just --quiet, False
and None arguments are missed, everything else is passed with str(value).

**Arguments:**

* `source`: html string to generate pdf from or url to get
* `quiet`: bool
* `grayscale`: bool
* `lowquality`: bool
* `margin_bottom`: string eg. 10mm
* `margin_left`: string eg. 10mm
* `margin_right`: string eg. 10mm
* `margin_top`: string eg. 10mm
* `orientation`: Portrait or Landscape
* `page_height`: string eg. 10mm
* `page_width`: string eg. 10mm
* `page_size`: string: A4, Letter, etc.
* `image_dpi`: int default 600
* `image_quality`: int default 94
* `extra_kwargs`: any exotic extra options for wkhtmltopdf

Returns string representing pdf

#### get_version()

Get version of pydf and wkhtmltopdf binary

#### get_help()

get help string from wkhtmltopdf binary
uses -h command line option

#### get_extended_help()

get extended help string from wkhtmltopdf binary
uses -H command line option

#### execute_wk(*args)

Low level function to call wkhtmltopdf, arguments are added to wkhtmltopdf binary and passed to subprocess with not processing.


HISTORY.rst
.. :changelog:

History
-------

v0.37.0 (2018-12-19)
....................
* upgrade wkhtmltopdf binary to "wkhtmltopdf 0.12.5 for Ubuntu 18.04 (bionic)", #11. Requires heroku-18

v0.36.0 (2017-06-13)
....................
* allow alternative wkhtmltopdf binary via ``WKHTMLTOPDF_PATH`` environment variable

v0.35.0 (2017-06-06)
....................
* add docker auto tagging

v0.34.0 (2017-06-06)
....................
* improve error output

v0.33.0 (2017-05-24)
....................
* remove pdf metadata modification as it can break some pdf viewers

v0.32.0 (2017-05-24)
....................
* set ``cache_dir`` for ``generate_pdf`` by default

v0.31.0 (2017-05-23)
....................
* move to python 3.6 +
* add ``async`` generation for roughly 10x speedup

v0.30.0
.......
* uprev wkhtmltopdf from **0.12.2 (beta)** to **0.12.4**.
* code cleanup
* this is the same as ``v0.3``, I made a mistake when versioning

v0.21.0
.......
* correct permissions on wkhtmltopdf binary.


tests/utils.py
from io import BytesIO, StringIO

import pdfminer.layout
from pdfminer import high_level


def pdf_text(pdf_data: bytes) -> str:
    laparams = pdfminer.layout.LAParams()
    output = StringIO()
    high_level.extract_text_to_fp(BytesIO(pdf_data), output, laparams=laparams)
    return output.getvalue()


tests/__init__.py
from .runtests import *

tests/runtests.py
# coding: utf8
import os
from unittest import TestCase
from pydf import generate_pdf, get_version, get_help, get_extended_help

import sys
if sys.version_info[0] == 3:
    def byteify(s):
        return bytes(s, encoding='utf8')
else:
    def byteify(s):
        return s


class PydfTestCase(TestCase):
    def test_generate_pdf_with_html(self):
        pdf_content = generate_pdf('<html><body>Is this thing on?</body></html>')
        assert pdf_content[:4] == byteify('%PDF')

    def test_generate_pdf_with_url(self):
        pdf_content = generate_pdf('http://google.com')
        assert pdf_content[:4] == byteify('%PDF')

    def test_unicode(self):
        pdf_content = generate_pdf(u'<html><body>Schrödinger</body></html>')
        assert pdf_content[:4] == byteify('%PDF')

    def test_extra_arguments(self):
        pdf_content = generate_pdf(
            '<html><body>testing</body></html>',
            quiet=False,
            grayscale=True,
            lowquality=True,
            margin_bottom='20mm',
            margin_left='20mm',
            margin_right='20mm',
            margin_top='20mm',
            orientation='Landscape',
            page_height=None,
            page_width=None,
            page_size='Letter',
            image_dpi='300',
            image_quality='70',
        )
        assert pdf_content[:4] == byteify('%PDF')

    def test_custom_size(self):
        pdf_content = generate_pdf(
            '<html><body>testing</body></html>',
            page_height='50mm',
            page_width='50mm',
        )
        assert pdf_content[:4] == byteify('%PDF')

    def test_extra_kwargs(self):
        pdf_content = generate_pdf(
            '<html><body>testing</body></html>',
            header_right='Page [page] of [toPage]'
        )
        assert pdf_content[:4] == byteify('%PDF')

    def test_wrong_path(self):
        os.environ['WKHTMLTOPDF_CMD'] = 'foo bar'
        self.assertRaises(IOError, get_help)
        del os.environ['WKHTMLTOPDF_CMD']

    def test_no_arguments(self):
        self.assertRaises(TypeError, generate_pdf)

    def test_bad_arguments(self):
        self.assertRaises(IOError, generate_pdf, 'www.')

    def test_get_version(self):
        print(get_version())

    def test_get_help(self):
        get_help()

    def test_get_extended_help(self):
        get_extended_help()


docker.py
#!/usr/bin/env python3.6
"""
pydf pdf generation

To generate PDF POST (or GET with data if possible) you HTML data to /generate.pdf

For example:

    docker run -rm -p 8000:80 -d samuelcolvin/pydf
    curl -d '<h1>this is html</h1>' http://localhost:8000/generate.pdf > created.pdf
    open "created.pdf"
"""

import os
from aiohttp import web
from pydf import AsyncPydf


async def index(request):
    return web.Response(text=__doc__)


async def generate(request):
    data = await request.read()
    if not data:
        raise web.HTTPBadRequest(text='400: no PDF data provided\n')
    pdf_content = await app['apydf'].generate_pdf(data.decode())
    return web.Response(body=pdf_content, content_type='application/pdf')

app = web.Application()
app.router.add_get('/', index)
app.router.add_route('*', '/generate.pdf', generate)
app['apydf'] = AsyncPydf(loop=app.loop)

port = int(os.getenv('PORT', '80'))
print(f'starting pydf server on port {port}', flush=True)
web.run_app(app, port=port, print=lambda v: None)


benchmark/run.py
import asyncio
from pathlib import Path
from time import time

from pydf import AsyncPydf, generate_pdf


THIS_DIR = Path(__file__).parent.resolve()
html = (THIS_DIR / 'invoice.html').read_text()
OUT_DIR = THIS_DIR / 'output'
if not OUT_DIR.exists():
    Path.mkdir(OUT_DIR)


def go_sync():
    count = 10
    for i in range(count):
        pdf = generate_pdf(
            html,
            page_size='A4',
            zoom='1.25',
            margin_left='8mm',
            margin_right='8mm',
        )
        print(f'{i:03}: {len(pdf)}')
        file = OUT_DIR / f'output_{i:03}.pdf'
        file.write_bytes(pdf)
    return count

start = time()
count = go_sync()
time_taken = (time() - start) / count
print(f'sync, time taken per pdf: {time_taken:0.3f}s')

async def go_async():
    apydf = AsyncPydf()

    async def gen(i_):
        pdf = await apydf.generate_pdf(
            html,
            page_size='A4',
            zoom='1.25',
            margin_left='8mm',
            margin_right='8mm',
        )
        print(f'{i_:03}: {len(pdf)}')
        f = OUT_DIR / f'output_{i_:03}.pdf'
        f.write_bytes(pdf)

    count = 20
    coros = map(gen, range(count))
    await asyncio.gather(*coros)
    return count


start = time()
loop = asyncio.get_event_loop()
count = loop.run_until_complete(go_async())
time_taken = (time() - start) / count
print(f'async time taken per pdf: {time_taken:0.3f}s')


docker-entrypoint.py
#!/usr/bin/env python3.6
"""
pydf

pdf generation in docker.

To generate PDF POST (or GET with data if possible) you HTML data to /generate.pdf.

Extra arguments can be passed using http headers; any header starting "pdf-" or "pdf_" will
have that prefix removed, be converted to lower case and passed to wkhtmltopdf.

For example:

    docker run -rm -p 8000:80 -d samuelcolvin/pydf
    curl -d '<h1>this is html</h1>' -H "pdf-orientation: landscape" http://localhost:8000/generate.pdf > created.pdf
    open "created.pdf"
"""
import os
import logging
from time import time

from aiohttp import web
from pydf import AsyncPydf

logger = logging.getLogger('main')
logger.setLevel(logging.INFO)
handler = logging.StreamHandler()
handler.setLevel(logging.INFO)
logger.addHandler(handler)


async def index(request):
    return web.Response(text=__doc__)


async def generate(request):
    start = time()
    config = {}
    for k, v in request.headers.items():
        if k.startswith('Pdf-') or k.startswith('Pdf_'):
            config[k[4:].lower()] = v.lower()
    data = await request.read()
    if not data:
        logger.info('Request with no body data')
        raise web.HTTPBadRequest(text='400: no HTML data to convert to PDF in request body\n')
    try:
        pdf_content = await app['apydf'].generate_pdf(data.decode(), **config)
    except RuntimeError as e:
        logger.info('Error generating PDF, time %0.2fs, config: %s', time() - start, config)
        return web.Response(text=str(e) + '\n', status=418)
    else:
        logger.info('PDF generated in %0.2fs, html-len %d, pdf-len %d', time() - start, len(data), len(pdf_content))
        return web.Response(body=pdf_content, content_type='application/pdf')

app = web.Application()
app.router.add_get('/', index)
app.router.add_route('*', '/generate.pdf', generate)
app['apydf'] = AsyncPydf(loop=app.loop)

port = int(os.getenv('PORT', '80'))
logger.info('starting pydf server on port %s', port)
web.run_app(app, port=port, print=lambda v: None)


pydf/wkhtmltopdf.py
import asyncio
import os
import subprocess
import tempfile
from pathlib import Path

from .version import VERSION

__all__ = [
    'AsyncPydf',
    'generate_pdf',
    'get_version',
    'get_help',
    'get_extended_help',
]

THIS_DIR = Path(__file__).parent.resolve()
WK_PATH = os.getenv('WKHTMLTOPDF_PATH', str(THIS_DIR / 'bin' / 'wkhtmltopdf'))
DFT_CACHE_DIR = Path(tempfile.gettempdir()) / 'pydf_cache'


def _execute_wk(*args, input=None):
    """
    Generate path for the wkhtmltopdf binary and execute command.

    :param args: args to pass straight to subprocess.Popen
    :return: stdout, stderr
    """
    wk_args = (WK_PATH,) + args
    return subprocess.run(wk_args, input=input, stdout=subprocess.PIPE, stderr=subprocess.PIPE)


def _convert_args(**py_args):
    cmd_args = []
    for name, value in py_args.items():
        if value in {None, False}:
            continue
        arg_name = '--' + name.replace('_', '-')
        if value is True:
            cmd_args.append(arg_name)
        else:
            cmd_args.extend([arg_name, str(value)])

    # read from stdin and write to stdout
    cmd_args.extend(['-', '-'])
    return cmd_args


class AsyncPydf:
    def __init__(self, *, max_processes=20, loop=None, cache_dir=DFT_CACHE_DIR):
        self.semaphore = asyncio.Semaphore(value=max_processes, loop=loop)
        self.loop = loop
        if not cache_dir.exists():
            Path.mkdir(cache_dir)
        self.cache_dir = cache_dir

    async def generate_pdf(self, html, **cmd_args):
        cmd_args = [WK_PATH] + _convert_args(cache_dir=self.cache_dir, **cmd_args)
        async with self.semaphore:
            p = await asyncio.create_subprocess_exec(
                *cmd_args,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                loop=self.loop
            )
            p.stdin.write(html.encode())
            p.stdin.close()
            await p.wait()
            pdf_content = await p.stdout.read()
            if p.returncode != 0 and pdf_content[:4] != b'%PDF':
                stderr = await p.stderr.read()
                raise RuntimeError('error running wkhtmltopdf, command: {!r}\n'
                                   'response: "{}"'.format(cmd_args, stderr.decode().strip()))
            return pdf_content


def generate_pdf(html, *,
                 cache_dir: Path=DFT_CACHE_DIR,
                 grayscale: bool=False,
                 lowquality: bool=False,
                 margin_bottom: str=None,
                 margin_left: str=None,
                 margin_right: str=None,
                 margin_top: str=None,
                 orientation: str=None,
                 page_height: str=None,
                 page_width: str=None,
                 page_size: str=None,
                 image_dpi: str=None,
                 image_quality: str=None,
                 **extra_kwargs):
    """
    Generate a pdf from either a url or a html string.

    After the html and url arguments all other arguments are
    passed straight to wkhtmltopdf

    For details on extra arguments see the output of get_help()
    and get_extended_help()

    All arguments whether specified or caught with extra_kwargs are converted
    to command line args with "'--' + original_name.replace('_', '-')"

    Arguments which are True are passed with no value eg. just --quiet, False
    and None arguments are missed, everything else is passed with str(value).

    :param html: html string to generate pdf from
    :param grayscale: bool
    :param lowquality: bool
    :param margin_bottom: string eg. 10mm
    :param margin_left: string eg. 10mm
    :param margin_right: string eg. 10mm
    :param margin_top: string eg. 10mm
    :param orientation: Portrait or Landscape
    :param page_height: string eg. 10mm
    :param page_width: string eg. 10mm
    :param page_size: string: A4, Letter, etc.
    :param image_dpi: int default 600
    :param image_quality: int default 94
    :param extra_kwargs: any exotic extra options for wkhtmltopdf
    :return: string representing pdf
    """
    if not cache_dir.exists():
        Path.mkdir(cache_dir)

    py_args = dict(
        cache_dir=cache_dir,
        grayscale=grayscale,
        lowquality=lowquality,
        margin_bottom=margin_bottom,
        margin_left=margin_left,
        margin_right=margin_right,
        margin_top=margin_top,
        orientation=orientation,
        page_height=page_height,
        page_width=page_width,
        page_size=page_size,
        image_dpi=image_dpi,
        image_quality=image_quality,
    )
    py_args.update(extra_kwargs)
    cmd_args = _convert_args(**py_args)

    p = _execute_wk(*cmd_args, input=html.encode())
    pdf_content = p.stdout

    # it seems wkhtmltopdf's error codes can be false, we'll ignore them if we
    # seem to have generated a pdf
    if p.returncode != 0 and pdf_content[:4] != b'%PDF':
        raise RuntimeError('error running wkhtmltopdf, command: {!r}\n'
                           'response: "{}"'.format(cmd_args, p.stderr.decode().strip()))
    return pdf_content


def _string_execute(*args):
    return _execute_wk(*args).stdout.decode().strip(' \n')


def get_version():
    """
    Get version of pydf and wkhtmltopdf binary

    :return: version string
    """
    try:
        wk_version = _string_execute('-V')
    except Exception as e:
        # we catch all errors here to make sure we get a version no matter what
        wk_version = '%s: %s' % (e.__class__.__name__, e)
    return 'pydf version: %s\nwkhtmltopdf version: %s' % (VERSION, wk_version)


def get_help():
    """
    get help string from wkhtmltopdf binary
    uses -h command line option

    :return: help string
    """
    return _string_execute('-h')


def get_extended_help():
    """
    get extended help string from wkhtmltopdf binary
    uses -H command line option

    :return: extended help string
    """
    return _string_execute('-H')
